src/Microsoft.Security.Utilities.Core/StringInput.cs

```diff
@@ -0,0 +1,141 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#nullable enable
+
+using System;
+
+#if NET
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.InteropServices;
+#endif
+using System.Text;
+
+namespace Microsoft.Security.Utilities;
+
+/// <summary>
+/// Internal data structure for textual input for secret masking/scaning. It is
+/// backed by <see cref="ReadOnlyMemory{Char}" on modern .NET and by <see
+/// cref="string"/> on legacy .NET Standard/Framework.
+/// </summary>
+internal readonly struct StringInput
+{
+    // Implementation notes:
+    //
+    // * Making this a ref struct and using ReadOnlySpan in place of
+    //   ReadOnlyMemory would be better, but that requires significant
+    //   additional refactoring due to pervasive use of iterators (yield
```

Comment:
are we going to change this? i.e., reconsider our use of iterators? or, assuming a far-flung future where we run only on modern .NET, etc., etc., will we want to retain this pattern? thoughts

Reply:
Yes, I'd like to change this and reconsider our use of iterators, but I wasn't sure it was the best use of time right now to dig in on that.

So I would:

1. Make StringInput ref struct and backed by Span on modern .NET (TODO: File issue, just saving my comment first)
2. If and when we drop .NET standard/Framework (🙏), delete StringInput and use ReadOnlySpan<char> directly everywhere.

Keeping StringInput internal means that we can do both without breaking changes. It will do no harm to have Memory<char> overload that calls .Span and forwards to Span overload so there won't be a need for any breaking changes to the above.